File: android/app/src/main/kotlin/com/thinkupllc/daystildisney/core/model/DisneyPark.kt
package com.thinkupllc.daystildisney.core.model

import androidx.compose.ui.graphics.Color

/**
 * All 12 Disney theme parks across all six resorts.
 * Each park carries its display metadata and color palette so the theme
 * engine can drive the full visual experience from a single enum value.
 */
enum class DisneyPark(
    val displayName: String,
    val resort: DisneyResort,
    /** Asset name used to load the castle/icon drawable. No extension — resolved at runtime. */
    val iconAssetName: String,
    val colorPalette: ParkColorPalette,
) {
    // ----- Walt Disney World -----

    MAGIC_KINGDOM(
        displayName = "Magic Kingdom",
        resort = DisneyResort.WALT_DISNEY_WORLD,
        iconAssetName = "castle_magic_kingdom",
        colorPalette = ParkColorPalette(
            primary = Color(0xFF1A3A6B),          // Royal blue
            secondary = Color(0xFFB8962E),         // Gold
            accent = Color(0xFFE8C547),            // Bright gold
            backgroundGradientStart = Color(0xFF0D1B3E),
            backgroundGradientEnd = Color(0xFF1A3A6B),
            textOnPrimary = Color.White,
        ),
    ),

    EPCOT(
        displayName = "EPCOT",
        resort = DisneyResort.WALT_DISNEY_WORLD,
        iconAssetName = "castle_epcot",
        colorPalette = ParkColorPalette(
            primary = Color(0xFF0077B6),           // EPCOT blue
            secondary = Color(0xFF48CAE4),         // Light teal
            accent = Color(0xFFADE8F4),            // Pale aqua
            backgroundGradientStart = Color(0xFF023E8A),
            backgroundGradientEnd = Color(0xFF0096C7),
            textOnPrimary = Color.White,
        ),
    ),

    HOLLYWOOD_STUDIOS(
        displayName = "Hollywood Studios",
        resort = DisneyResort.WALT_DISNEY_WORLD,
        iconAssetName = "castle_hollywood_studios",
        colorPalette = ParkColorPalette(
            primary = Color(0xFF8B1A1A),           // Deep red
            secondary = Color(0xFFD4A017),         // Hollywood gold
            accent = Color(0xFFFF6B35),            // Warm orange
            backgroundGradientStart = Color(0xFF2C0A0A),
            backgroundGradientEnd = Color(0xFF8B1A1A),
            textOnPrimary = Color.White,
        ),
    ),

    ANIMAL_KINGDOM(
        displayName = "Animal Kingdom",
        resort = DisneyResort.WALT_DISNEY_WORLD,
        iconAssetName = "castle_animal_kingdom",
        colorPalette = ParkColorPalette(
            primary = Color(0xFF2D6A4F),           // Forest green
            secondary = Color(0xFF8B5E3C),         // Earth brown
            accent = Color(0xFF95D5B2),            // Sage green
            backgroundGradientStart = Color(0xFF1B4332),
            backgroundGradientEnd = Color(0xFF2D6A4F),
            textOnPrimary = Color.White,
        ),
    ),

    // ----- Disneyland Resort -----

    DISNEYLAND(
        displayName = "Disneyland",
        resort = DisneyResort.DISNEYLAND_RESORT,
        iconAssetName = "castle_disneyland",
        colorPalette = ParkColorPalette(
            primary = Color(0xFF8E44AD),           // Purple
            secondary = Color(0xFFF1A7D0),         // Pink
            accent = Color(0xFFFFD700),            // Gold
            backgroundGradientStart = Color(0xFF4A235A),
            backgroundGradientEnd = Color(0xFF8E44AD),
            textOnPrimary = Color.White,
        ),
    ),

    CALIFORNIA_ADVENTURE(
        displayName = "Disney California Adventure",
        resort = DisneyResort.DISNEYLAND_RESORT,
        iconAssetName = "castle_california_adventure",
        colorPalette = ParkColorPalette(
            primary = Color(0xFFD35400),           // Warm orange
            secondary = Color(0xFF2980B9),         // Sky blue
            accent = Color(0xFFF39C12),            // Sunburst yellow
            backgroundGradientStart = Color(0xFF7D3C00),
            backgroundGradientEnd = Color(0xFFD35400),
            textOnPrimary = Color.White,
        ),
    ),

    // ----- Tokyo Disney Resort -----

    TOKYO_DISNEYLAND(
        displayName = "Tokyo Disneyland",
        resort = DisneyResort.TOKYO_DISNEY_RESORT,
        iconAssetName = "castle_tokyo_disneyland",
        colorPalette = ParkColorPalette(
            primary = Color(0xFFCB4154),           // Cherry blossom red
            secondary = Color(0xFFFFC5C5),         // Sakura pink
            accent = Color(0xFFFFE4E1),            // Pale blush
            backgroundGradientStart = Color(0xFF7B1530),
            backgroundGradientEnd = Color(0xFFCB4154),
            textOnPrimary = Color.White,
        ),
    ),

    TOKYO_DISNEYSEA(
        displayName = "Tokyo DisneySea",
        resort = DisneyResort.TOKYO_DISNEY_RESORT,
        iconAssetName = "castle_tokyo_disneysea",
        colorPalette = ParkColorPalette(
            primary = Color(0xFF1B3A5C),           // Deep ocean blue
            secondary = Color(0xFF2E86AB),         // Mediterranean blue
            accent = Color(0xFF48CAE4),            // Aqua
            backgroundGradientStart = Color(0xFF0A1628),
            backgroundGradientEnd = Color(0xFF1B3A5C),
            textOnPrimary = Color.White,
        ),
    ),

    // ----- Disneyland Paris -----

    DISNEYLAND_PARK_PARIS(
        displayName = "Disneyland Park Paris",
        resort = DisneyResort.DISNEYLAND_PARIS,
        iconAssetName = "castle_disneyland_paris",
        colorPalette = ParkColorPalette(
            primary = Color(0xFF9B59B6),           // Lavender purple
            secondary = Color(0xFFE8A0BF),         // Rose pink
            accent = Color(0xFFE8D5F5),            // Soft violet
            backgroundGradientStart = Color(0xFF4C1B6B),
            backgroundGradientEnd = Color(0xFF9B59B6),
            textOnPrimary = Color.White,
        ),
    ),

    WALT_DISNEY_STUDIOS_PARK(
        displayName = "Walt Disney Studios Park",
        resort = DisneyResort.DISNEYLAND_PARIS,
        iconAssetName = "castle_wds_paris",
        colorPalette = ParkColorPalette(
            primary = Color(0xFF2C3E50),           // Dark slate
            secondary = Color(0xFFE74C3C),         // Studio red
            accent = Color(0xFFF39C12),            // Spotlight yellow
            backgroundGradientStart = Color(0xFF1A252F),
            backgroundGradientEnd = Color(0xFF2C3E50),
            textOnPrimary = Color.White,
        ),
    ),

    // ----- Hong Kong Disneyland -----

    HONG_KONG_DISNEYLAND_PARK(
        displayName = "Hong Kong Disneyland",
        resort = DisneyResort.HONG_KONG_DISNEYLAND,
        iconAssetName = "castle_hong_kong",
        colorPalette = ParkColorPalette(
            primary = Color(0xFF008080),           // Teal
            secondary = Color(0xFFB8962E),         // Gold
            accent = Color(0xFF4ECDC4),            // Bright teal
            backgroundGradientStart = Color(0xFF004040),
            backgroundGradientEnd = Color(0xFF008080),
            textOnPrimary = Color.White,
        ),
    ),

    // ----- Shanghai Disneyland -----

    SHANGHAI_DISNEYLAND_PARK(
        displayName = "Shanghai Disneyland",
        resort = DisneyResort.SHANGHAI_DISNEYLAND,
        iconAssetName = "castle_shanghai",
        colorPalette = ParkColorPalette(
            primary = Color(0xFF1A237E),           // Sapphire blue
            secondary = Color(0xFF9E9E9E),         // Silver
            accent = Color(0xFF90CAF9),            // Light blue
            backgroundGradientStart = Color(0xFF0D1340),
            backgroundGradientEnd = Color(0xFF1A237E),
            textOnPrimary = Color.White,
        ),
    );
}

File: ios/Days-Til-Disney/Days-Til-Disney/Core/Models/DisneyPark.swift
import Foundation

/// All 12 Disney theme parks across the six global resorts.
enum DisneyPark: String, Codable, CaseIterable, Identifiable {
    case magicKingdom
    case epcot
    case hollywoodStudios
    case animalKingdom
    case disneyland
    case californiaAdventure
    case tokyoDisneyland
    case tokyoDisneySea
    case disneylandParkParis
    case waltDisneyStudiosPark
    case hongKongDisneylandPark
    case shanghaiDisneylandPark

    var id: String { rawValue }

    var displayName: String {
        switch self {
        case .magicKingdom:          return "Magic Kingdom"
        case .epcot:                 return "EPCOT"
        case .hollywoodStudios:      return "Hollywood Studios"
        case .animalKingdom:         return "Animal Kingdom"
        case .disneyland:            return "Disneyland"
        case .californiaAdventure:   return "Disney California Adventure"
        case .tokyoDisneyland:       return "Tokyo Disneyland"
        case .tokyoDisneySea:        return "Tokyo DisneySea"
        case .disneylandParkParis:   return "Disneyland Park Paris"
        case .waltDisneyStudiosPark: return "Walt Disney Studios Park"
        case .hongKongDisneylandPark: return "Hong Kong Disneyland"
        case .shanghaiDisneylandPark: return "Shanghai Disneyland"
        }
    }

    var resort: DisneyResort {
        switch self {
        case .magicKingdom, .epcot, .hollywoodStudios, .animalKingdom:
            return .waltDisneyWorld
        case .disneyland, .californiaAdventure:
            return .disneylandResort
        case .tokyoDisneyland, .tokyoDisneySea:
            return .tokyoDisneyResort
        case .disneylandParkParis, .waltDisneyStudiosPark:
            return .disneylandParis
        case .hongKongDisneylandPark:
            return .hongKongDisneyland
        case .shanghaiDisneylandPark:
            return .shanghaiDisneyland
        }
    }

    var colorPalette: ParkColorPalette {
        switch self {
        case .magicKingdom:          return .magicKingdom
        case .epcot:                 return .epcot
        case .hollywoodStudios:      return .hollywoodStudios
        case .animalKingdom:         return .animalKingdom
        case .disneyland:            return .disneyland
        case .californiaAdventure:   return .californiaAdventure
        case .tokyoDisneyland:       return .tokyoDisneyland
        case .tokyoDisneySea:        return .tokyoDisneySea
        case .disneylandParkParis:   return .disneylandParkParis
        case .waltDisneyStudiosPark: return .waltDisneyStudiosPark
        case .hongKongDisneylandPark: return .hongKongDisneylandPark
        case .shanghaiDisneylandPark: return .shanghaiDisneylandPark
        }
    }

    var iconAssetName: String {
        switch self {
        case .magicKingdom:          return "icon-magic-kingdom"
        case .epcot:                 return "icon-epcot"
        case .hollywoodStudios:      return "icon-hollywood-studios"
        case .animalKingdom:         return "icon-animal-kingdom"
        case .disneyland:            return "icon-disneyland"
        case .californiaAdventure:   return "icon-california-adventure"
        case .tokyoDisneyland:       return "icon-tokyo-disneyland"
        case .tokyoDisneySea:        return "icon-tokyo-disneysea"
        case .disneylandParkParis:   return "icon-paris-disneyland"
        case .waltDisneyStudiosPark: return "icon-paris-studios"
        case .hongKongDisneylandPark: return "icon-hong-kong"
        case .shanghaiDisneylandPark: return "icon-shanghai"
        }
    }

    /// The UUID used by the ThemeParks.wiki API to identify this park.
    /// Returns nil for parks that are not currently mapped (should not occur for any
    /// of the 12 supported parks, but callers must handle the optional gracefully).
    var themeParkWikiID: String? {
        switch self {
        case .magicKingdom:          return "75ea578a-adc8-4116-a54d-dccb60765ef9"
        case .epcot:                 return "47f90d2c-e191-4239-a466-5571c099571a"
        case .hollywoodStudios:      return "288747d1-8b4f-4a64-867e-ea7c9b27f573"
        case .animalKingdom:         return "1c84a229-8862-4648-9c71-378ddd2c7693"
        case .disneyland:            return "7340550b-c14d-4def-80bb-acdb51d49a66"
        case .californiaAdventure:   return "832fcd51-ea19-4e77-85c7-75d5843b127c"
        case .disneylandParkParis:   return "ca888437-ebb4-4d50-aed2-d227f7096968"
        case .waltDisneyStudiosPark: return "fc1ecaae-8af4-482b-af05-e0b9b0e7c3a6"
        case .tokyoDisneyland:       return "67b290d5-3478-4e5f-adb4-b99c42d68bfe"
        case .tokyoDisneySea:        return "8b4de4d1-b68c-431a-b83d-6f7c6fa949c2"
        case .shanghaiDisneylandPark: return "dae968d5-630d-4719-8b06-3d107e944401"
        case .hongKongDisneylandPark: return "bd0eb47b-2f02-4d4d-90fa-cb3a68988e3b"
        }
    }

    var emoji: String {
        switch self {
        case .magicKingdom:          return "🏰"
        case .epcot:                 return "🌍"
        case .hollywoodStudios:      return "🎬"
        case .animalKingdom:         return "🦁"
        case .disneyland:            return "🏰"
        case .californiaAdventure:   return "🌊"
        case .tokyoDisneyland:       return "🌸"
        case .tokyoDisneySea:        return "⛵"
        case .disneylandParkParis:   return "🗼"
        case .waltDisneyStudiosPark: return "🎥"
        case .hongKongDisneylandPark: return "✨"
        case .shanghaiDisneylandPark: return "🌟"
        }
    }
}
